platform/default/bidi.cpp

```diff
@@ -100,6 +100,108 @@ std::vector<std::u16string> BiDi::processText(const std::u16string& input,
 
     return applyLineBreaking(lineBreakPoints);
 }
+    
+std::vector<StyledText> BiDi::processStyledText(const StyledText& input, std::set<std::size_t> lineBreakPoints) {
```

Comment:
It would be pretty straightforward to add unit tests that were basically a native port of https://github.com/mapbox/mapbox-gl-rtl-text/blob/master/test/arabic.test.js, but this functionality is already exercised pretty directly by the `formatted/arabic` render test. When we can test effectively either way, I generally prefer render tests because they introduce less friction... what do you think?

Reply:
Added the test -- and it turned something up that render tests couldn't see! Lines that had characters removed from them during BiDi due to UBIDI_REMOVE_BIDI_CONTROLS could end up with extra null characters at the trailing end of the string (I don't think this would actually affect rendering, but it wasn't intentional and it'd be a surprise for anyone consuming these strings). I changed `BiDi::getLine` to match the behavior of `BiDi::writeReverse` and explicitly resize the output string before returning it.

Reply:
Even more interesting: the sanitizer found a problem in the ICU code?

`../../../vendor/icu/src/ubidiln.cpp:668:17: runtime error: left shift of 2 by 31 places cannot be represented in type 'int32_t' (aka 'int')`

Reply:
The sanitizer issue looks to me like it's legitimately undefined behavior in the C++ spec, although I have no idea if there are any compilers that would fail to get the last bit out the way the two problematic macros are trying to do. I looked on the ICU issue tracker and couldn't find an open issue, we should probably file one?

I tried applying a patch in our vendoring code, based on a patch used by LibreOffice: https://github.com/LibreOffice/core/blob/master/external/icu/icu4c-ubsan.patch.1 This is pretty brittle -- @kkaefer do you think it's reasonable until there's an upstream fix?

Reply:
Yeah, this fix looks good; let's apply it until there's an upstream fix.